### PR TITLE
feat: add tycho-indexer metrics 

### PR DESCRIPTION
**Pull request by louise-poole**

### CODE REVIEW

File: tycho-indexer/src/substreams/stream.rs [hunk 1]
```diff
@@ -98,6 +98,14 @@ fn stream_blocks(
                     for await response in stream {
                         match process_substreams_response(response).await {
                             BlockProcessedResult::BlockScopedData(block_scoped_data) => {
+                                let time_ms = SystemTime::now()
+                                    .duration_since(UNIX_EPOCH)
+                                    .expect("Time went backwards")
+                                    .as_millis();
+
+                                gauge!("substreams_block_number").set(block_scoped_data.final_block_height as f64);
```
Comment (kayibal) [inline tycho-indexer/src/substreams/stream.rs]: This is the finalized block number, not the current blocks number.

File: tycho-indexer/src/substreams/stream.rs [hunk 2]
```diff
@@ -98,6 +98,14 @@ fn stream_blocks(
...
+                                gauge!("substreams_block_timestamp").set(time_ms as f64);
```
Comment (kayibal) [inline tycho-indexer/src/substreams/stream.rs]: Confusing name imo, this should be `substreams_block_witnessed_ts` or similar

File: tycho-indexer/src/substreams/stream.rs [hunk 3]
```diff
@@ -103,6 +104,14 @@ fn stream_blocks(
...
                         match process_substreams_response(response).await {
                             BlockProcessedResult::BlockScopedData(block_scoped_data) => {
+                                if let Some(block) = block_scoped_data.clock.clone() {
+                                    if let Some(block_ts) = block.timestamp {
+                                        let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time went backwards!?").as_secs();
+                                        let lag = max(now - block_ts.seconds as u64, 0);
+                                        gauge!("substreams_lag_seconds", "extractor" => extractor_id.clone()).set(lag as f64);
+                                    }
+                                };
+
```
Comment (kayibal) [inline tycho-indexer/src/substreams/stream.rs:114]: Yeah much simpler! A bit surprised that these fields are optional though. Any idea why?

File: tycho-core/src/dto.rs [hunk 4]
```diff
@@ -208,7 +208,6 @@ pub struct Transaction {
 }
 
 impl Transaction {
-    #[allow(clippy::too_many_arguments)]
```
Comment (tvinagre) [inline tycho-core/src/dto.rs:211]: Unrelated to the PR but ok

File: tycho-indexer/src/extractor/protocol_extractor.rs [hunk 5]
```diff
@@ -190,6 +192,11 @@ where
             );
             state.last_report_ts = now;
             state.last_report_block_number = block.number;
+            gauge!(
+                "extractor_sync_time_remaining",
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/protocol_extractor.rs]: In this case I think it's nice to add the unit
```suggestion
                "extractor_sync_remaining_minutes",
```

File: tycho-indexer/src/extractor/runner.rs [hunk 6]
```diff
@@ -173,6 +175,11 @@ impl ExtractorRunner {
                             Some(Ok(BlockResponse::New(data))) => {
                                 let block_number = data.clock.as_ref().map(|v| v.number).unwrap_or(0);
                                 tracing::Span::current().record("block_number", block_number);
+                                gauge!("extractor_current_block", "extractor_id" => id.to_string()).set(block_number as f64);
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/runner.rs]:
```suggestion
                                gauge!("extractor_current_block_number", "extractor_id" => id.to_string()).set(block_number as f64);
```

File: tycho-indexer/src/extractor/runner.rs [hunk 7]
```diff
@@ -188,6 +195,9 @@ impl ExtractorRunner {
                                         return Err(err);
                                     }
                                 }
+
+                                let duration = start_time.elapsed();
+                                histogram!("block_processing_time").record(duration.as_millis() as f64);
```
Comment (tvinagre) [inline tycho-indexer/src/extractor/runner.rs]:
```suggestion
                                histogram!("block_processing_time_ms").record(duration.as_millis() as f64);
```

File: tycho-indexer/src/main.rs [hunk 8]
```diff
@@ -118,6 +119,11 @@ fn create_tracing_subscriber() {
 
 /// Creates and runs the Prometheus metrics exporter using Actix Web.
 pub fn create_metrics_exporter() -> tokio::task::JoinHandle<()> {
+    describe_gauge!(
+        "extractor_sync_time_remaining",
+        "Time remaining for the extractor to sync, measured in minutes."
```
Comment (tvinagre) [inline tycho-indexer/src/main.rs]: Ah, you did describe them here :)
Does this appear on Grafana when selecting VictoriaMetrics metrics?

File: tycho-indexer/src/services/rpc.rs [hunk 9]
```diff
@@ -6,10 +6,18 @@ use actix_web::{web, HttpResponse, ResponseError};
 use anyhow::Error;
 use chrono::{Duration, Utc};
 use diesel_async::pooled_connection::deadpool;
+use metrics::counter;
 use reqwest::StatusCode;
 use thiserror::Error;
 use tracing::{debug, error, info, instrument, trace, warn};
 
+use tycho_core::{
```
Comment (tvinagre) [inline tycho-indexer/src/services/rpc.rs:14]: A lot of formatting changes - I'm assuming this isn't being covered by CI?

File: tycho-indexer/src/substreams/stream.rs [hunk 10]
```diff
@@ -144,7 +163,7 @@ fn stream_blocks(
                     // We failed to connect and will try again; this is another
                     // case where we actually _want_ to back off in case we keep
                     // having connection errors.
-
+                    counter!("substreams_failure", "module" => output_module_name.clone(), "cause" => "connection_error").increment(1);
```
Comment (tvinagre) [inline tycho-indexer/src/substreams/stream.rs:166]: Some metrics you add a "total" in the end, like (rpc_requests_total) - and other you leave it out - is there a reason? I'd like to have a pattern here

**Review by tvinagre — APPROVED** [review]
Looking good! Smol comments on metric names

File: tycho-indexer/src/substreams/stream.rs (hunk 3 above)
Comment (louise-poole) [reply tycho-indexer/src/substreams/stream.rs:114]: No clue - I ran this a few times and had a test metric to log whenever it was None and it was never logged. It seems like in general we can expect a value here.

File: tycho-core/src/dto.rs (hunk 4 above)
Comment (louise-poole) [reply tycho-core/src/dto.rs:211]: Got distracted while we were all debating what path to take for the lag metric 😆  I added this skip somewhere else within the PR and noticed it was used a lot where it wasn't needed.

File: tycho-indexer/src/main.rs (hunk 8 above)
Comment (louise-poole) [reply tycho-indexer/src/main.rs]: According to the docs, this should appear. In my local test I haven't seen it though. I think I will rather rename the metrics as you've suggested to be 100% clear.

File: tycho-indexer/src/substreams/stream.rs (hunk 10 above)
Comment (louise-poole) [reply tycho-indexer/src/substreams/stream.rs:166]: I actually started to remove them from the metric names as it was making them long without adding any info. I missed this one I suppose.

File: tycho-indexer/src/services/rpc.rs (hunk 9 above)
Comment (louise-poole) [reply tycho-indexer/src/services/rpc.rs:14]: No - for the most part our code follows the pattern of: std imports, external imports, repo imports, crate imports. If I worked on a file that broke this pattern I fixed it, but these are inconsequential changes only done for consistency.

**Pull request merged**